### PR TITLE
fixed #577: taking message type into account when merging messages

### LINKED ISSUE #577: taking message type into account when merging messages
The merge operation on Navajo message currently does not correctly take the message type into account. Consider this example:
```xml
<message name="Source">
  <message name="M1" type="array">
    <message name="M1" type="definition">
       <property name="P1" />
    </message>
  </message>
  <message name="M2" type="simple">
    <property name="P2"/>
  </message>
</message>

<message name="Target">
  <message name="M1" type="simple">
    <property name="P1"/>
  </message>
  <message name="M2" type="array">
    <message name="M2" type="definition">
       <property name="P2" />
    </message>
  </message>
</message>
```

If these two messages will be merged: ```source.merge(target)```, the current result will be:
```xml
<message name="Source">
  <message name="M1" type="simple">
    <property name="P1"/>
  </message>
  <message name="M2" type="array">
    <message name="M2" type="definition">
       <property name="P2" />
    </message>
  </message>
</message>
```
Note that the message type for M1 and M2 have changed. This should not happen, especially in combination with entities where an entity definition may define a message as simple, and the output of the response can change its type to array (or the other way around), and yet the client relies on the entity definition. 

Therefore incompatible message types should not be considered mergeable and an exception should be thrown

### PR DESCRIPTION
**Pull request by roelofkemp**
see: https://github.com/Dexels/navajo/issues/577

### CODE REVIEW

**Comment by kharybdys** [comment]
See comment in issue.

**Comment by kharybdys** [comment]
Also your fix should probably include a check on preferThis -> depending on its value prefer the one or the other submessage structure found

**Comment by roelofkemp** [comment]
Discussed with Marte, we prefer to throw when incompatible messages are merged.

**Review by kharybdys — APPROVED** [review]

File: core/com.dexels.navajo.document/src/com/dexels/navajo/document/base/BaseMessageImpl.java [hunk 1]
```diff
@@ -1636,6 +1636,9 @@ public void merge(Message incoming, boolean preferThis) {
 
     @Override
     public void merge(Message incoming, boolean preferThis, boolean applySubType) {
+        if (isArrayMessage() != incoming.isArrayMessage()) {
+            throw new NavajoException("Incompatible message types in merge: " + getFullMessageName() + " (" + getType() +  "), " + incoming.getFullMessageName() + " (" + incoming.getType() +  ")");
```
Comment (gballintijn) [inline core/com.dexels.navajo.document/src/com/dexels/navajo/document/base/BaseMessageImpl.java:1640]: Plz wrap this line.

File: core/com.dexels.navajo.document/src/com/dexels/navajo/document/base/BaseMessageImpl.java [hunk 2]
```diff
@@ -1671,37 +1674,31 @@ public void merge(Message incoming, boolean preferThis, boolean applySubType) {
...
+        for (int i = 0; i < incomingSubMessages.size(); i++) {
+            Message incomingSubMessage = incomingSubMessages.get(i);
+            // check if we have a sub message with the same name
+            Message existingSubMessage = getMessage(incomingSubMessage.getName());
+            if (existingSubMessage != null) {
+                // check if the existing sub message is of the same type (array/simple) as the
+                // incoming message, if not, we cannot use it
+                if (existingSubMessage.isArrayMessage() != incomingSubMessage.isArrayMessage()) {
+                    throw new NavajoException("Incompatible message types in merge: " + existingSubMessage.getFullMessageName() + " (" + existingSubMessage.getType() +  "), " + incomingSubMessage.getFullMessageName() + " (" + incomingSubMessage.getType() +  ")");
+                }
```
Comment (gballintijn) [inline core/com.dexels.navajo.document/src/com/dexels/navajo/document/base/BaseMessageImpl.java:1687]: Plz wrap this line.

**Comment by gballintijn** [comment]
Approved based on limited knowledge.

**Review by gballintijn — APPROVED** [review]

**Pull request merged**